Enhance Dynalite Integration after review

ziv1234 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/12157

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ziv1234 committed:
fixes per Martin Hjelmare

ziv1234 committed:
pylint fix

ziv1234 committed:
final fixes per request

ziv1234 committed:
fixed unit tests for new config flow

ziv1234 committed:
Added unit-tests to increase coverage. at 97% now

ziv1234 commented:
all tests pass now include coverage. Should be ready to merge as this is just a code quality change. No functionality is affected

ziv1234 committed:
Added unit tests for 100% coverage of component

MartinHjelmare commented on tests/components/dynalite/test_config_flow.py:
```diff
@@ -3,33 +3,27 @@
+    entry = Mock()
+    host = "abc"
+    entry.data = {"host": host}
+    hass.config_entries.async_entries = Mock(return_value=[entry])
```
Don't patch the core. Use `MockConfigEntry` from tests/common.py and add a mock entry to hass. Then start the flow via core `hass.config_entries.flow.async_init`. We can probably remove `configured_hosts` helper when using the unique_id config flow api.

ziv1234 replied:
done

MartinHjelmare commented on tests/components/dynalite/test_config_flow.py:
```diff
@@ -3,33 +3,27 @@
     flow_handler = config_flow.DynaliteFlowHandler()
     with patch.object(flow_handler, "context", create=True):
-        with patch.object(flow_handler, "hass", create=True) as mock_hass:
+        with patch.object(flow_handler, "hass", create=True):
```
We have `hass` available in all tests as a pytest fixture. Please see existing config flow tests for best practices. The hue config flow tests are good examples.

ziv1234 replied:
done

MartinHjelmare commented on tests/components/dynalite/test_init.py:
```diff
@@ -11,6 +11,21 @@
+    config = {}
+    mock_conf_host = Mock(return_value=[])
+    with patch(
+        "homeassistant.components.dynalite.__init__.configured_hosts", mock_conf_host
```
This path is incorrect. The `__init__.py` module is the dynalite package and is not written out in the path.

ziv1234 replied:
fixed

MartinHjelmare commented on tests/components/dynalite/test_init.py:
```diff
@@ -11,6 +11,21 @@
+    with patch(
+        "homeassistant.components.dynalite.__init__.configured_hosts", mock_conf_host
+    ):
+        assert await async_setup(hass, config)
```
Don't call integration methods directly. Use the core to set up the tests and then assert the state from the core interfaces. Eg if we want to assert the state of an entity we get the state from the core state machine, `hass.states`.

Use `async_setup_component` from homeassistant/setup.py, to set up the integration.

ziv1234 replied:
fixed

MartinHjelmare commented on tests/components/dynalite/test_light.py:
```diff
@@ -15,8 +16,13 @@
     await async_setup_entry(hass, entry, async_add)
-    bridge.register_add_entities.assert_called_once()
-    assert bridge.register_add_entities.mock_calls[0] == call(async_add)
+    bridge.register_add_devices.assert_called_once()
```
In the light tests we just want to assert the state of the added light entity and that we can call the light services and that those calls call into the correct dynalite api. Patch the dynalite library methods and assert the calls of those patched methods.

Set up the dynalite component, which will load the light platform, which will add an entity. Then do the assertions.

ziv1234 replied:
fixed

MartinHjelmare commented on tests/components/dynalite/test_light.py:
```diff
@@ -38,7 +44,38 @@
+    """Test supported feaures didn't change."""
+    device = Mock()
+    bridge = Mock()
+    dyn_light = DynaliteLight(device, bridge)
```
Don't create the entity directly. Set up the component and platform which will add the entity. Then assert the state attributes via the core state machine.

ziv1234 replied:
fixed

MartinHjelmare reviewed (changes requested):
Please always use the core interfaces in the tests, and never the integration interfaces directly. Doing this will make the tests more robust, and we'll be able to change the implementation without changing the tests as long as the result in the core remains the same.

MartinHjelmare commented on homeassistant/components/dynalite/light.py:
```diff
@@ -77,8 +88,15 @@ def supported_features(self):
+        async_dispatcher_connect(
+            self.hass,
+            self._bridge.update_signal(self._device),
+            self.schedule_update_ha_state,
```
```suggestion
            self.async_schedule_update_ha_state,
```

MartinHjelmare commented on homeassistant/components/dynalite/light.py:
```diff
@@ -77,8 +88,15 @@ def supported_features(self):
+        )
+        # register for wide update
+        async_dispatcher_connect(
+            self.hass, self._bridge.update_signal(), self.schedule_update_ha_state
```
```suggestion
            self.hass, self._bridge.update_signal(), self.async_schedule_update_ha_state
```

MartinHjelmare commented on homeassistant/components/dynalite/bridge.py:
```diff
@@ -56,54 +43,42 @@ def __init__(self, hass, config_entry):
-        """Add the entities to HA if async_add_entities was registered, otherwise queue until it is."""
-        if not entities:
+    def add_devices_when_registered(self, devices):
+        """Add the devices to HA if async_add_entities was registered, otherwise queue until it is."""
```
```suggestion
        """Add the devices to HA if the add devices callback was registered, otherwise queue until it is."""
```

MartinHjelmare commented on homeassistant/components/dynalite/config_flow.py:
```diff
@@ -32,27 +30,12 @@ def __init__(self):
         """Import a new bridge as a config entry."""
         LOGGER.debug("async_step_import - %s", import_info)
         host = self.context[CONF_HOST] = import_info[CONF_HOST]
```
Where do we use the context?

ziv1234 replied:
I am quite a newbie in these config flows. I just copied it from an example. I don't use it explicitly. Can I just remove this part?

MartinHjelmare replied:
Yes. It looks like we can remove it.

ziv1234 replied:
removed

ziv1234 committed:
removed configured_host function and updated config_flow unit tests

ziv1234 committed:
added a pylint directive since it tells me by mistake DOMAIN is not used

ziv1234 committed:
fixed path (removed __init__)

ziv1234 committed:
Update homeassistant/components/dynalite/light.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
Update homeassistant/components/dynalite/light.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
fixed the test as we moved from schedule_update_... to async_schedule

ziv1234 committed:
Update homeassistant/components/dynalite/bridge.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
removed context from config_flow
changed test_init to use the core methods

ziv1234 committed:
Merge branch 'dev' of https://github.com/ziv1234/home-assistant into dev

MartinHjelmare commented on homeassistant/components/dynalite/config_flow.py:
```diff
@@ -31,28 +20,13 @@ def __init__(self):
+        host = import_info[CONF_HOST]
+        await self.async_set_unique_id(host)
+        self._abort_if_unique_id_configured()
         return await self._entry_from_bridge(host)
```
We want to test that the connection to the bridge works for this host, before creating the entry.

ziv1234 replied:
this will be quite tricky since if I create the bridge here, I cannot pass it to the async_setup_entry. I looked in the hue component that is similar, and they create a bridge to try and connect and then another one when it is set up, which I can also do, although it will include some code duplication. 

Is this something that would be helpful? As this is coming from static configuration.yaml, it may be cleaner to leave as is. Your call

ziv1234 replied:
ok. I did something. initially it works by polling (every second for 30 seconds then times out). I figured it could only cause a short delay the first time we configure, so it should be ok to any user

MartinHjelmare replied:
We want to create a bridge here and then a new bridge in `async_setup_entry`. That's our standard way.

ziv1234 replied:
ok. this is what i did (hopefully...)

MartinHjelmare commented on tests/components/dynalite/test_init.py:
```diff
@@ -1,74 +1,91 @@
-            old_host: {"host": old_host},
-        }
-        hass.async_create_task.assert_called_once()
+from tests.common import MockConfigEntry, mock_coro
```
We don't need `mock_coro` when we use `asynctest.patch`. It will patch coroutine functions with `CoroutineMock` and normal functions with `MagicMock`.

ziv1234 replied:
done. now only using it when assigning return_values in a mock

MartinHjelmare replied:
For those cases we should use `asynctest.CoroutineMock` instead.

ziv1234 replied:
fixed. shouldn't have any more mock_coro in the code

MartinHjelmare commented on tests/components/dynalite/test_init.py:
```diff
@@ -1,74 +1,91 @@
+    """Test a successful setup."""
+    host = "1.2.3.4"
+    port = 789
+    with patch.object(
```
This should be enough:
```py
with patch("homeassistant.components.dynalite.DynaliteBridge.async_setup", return_value=True):
```

Note that the bridge setup function never returns anything else than `True` which seems to be a bug. We should probably return `False` if we fail to set up the bridge connection.

ziv1234 replied:
the connection is happening asynchronously in the library, so will be quite difficult to test when we set it up. Ok if I defer this to later?

MartinHjelmare replied:
We should at least clean up the code that pretends to check for a valid setup.

ziv1234 replied:
fixed

MartinHjelmare commented on tests/components/dynalite/test_init.py:
```diff
@@ -1,74 +1,91 @@
-    """Test setup of an entry."""
+async def test_empty_config(hass):
+    """Test with an empty config."""
+    assert await async_setup_component(hass, dynalite.DOMAIN, {}) is True
```
We can remove `is True`.

ziv1234 replied:
done. i copied from the hue platform where it also exists, but agree it is redundant

MartinHjelmare commented on tests/components/dynalite/test_init.py:
```diff
@@ -1,74 +1,91 @@
-        async_mock(mock_async_setup),
+    port = 789
+    with patch.object(
+        dynalite.DynaliteBridge, "async_setup", return_value=mock_coro(False)
```
This can currently not happen. We should fix that.

ziv1234 committed:
moved test_light to also use the core interfaces

ziv1234 committed:
moved to config_entries.async_unload

ziv1234 commented on tests/components/dynalite/test_init.py:
```diff
@@ -1,101 +1,91 @@
+    assert hass.data[dynalite.DOMAIN].get(entry.entry_id)
+
+    mock_bridge.return_value.async_reset.return_value = mock_coro(True)
+    assert await dynalite.async_unload_entry(hass, entry)
```
fixed

ziv1234 committed:
additional fixes for the tests

MartinHjelmare commented on tests/components/dynalite/test_light.py:
```diff
@@ -1,44 +1,86 @@
+    """Set up the component and platform and create a light based on the device provided."""
+    host = "1.2.3.4"
+    with patch.object(
+        dynalite.bridge.DynaliteDevices, "async_setup", return_value=True
```
```suggestion
        "homeassistant.components.dynalite.bridge.DynaliteDevices.async_setup", return_value=True
```

MartinHjelmare commented on tests/components/dynalite/test_light.py:
```diff
@@ -1,44 +1,86 @@
+                        }
+                    },
+                )
+                is True
```
```suggestion
```

MartinHjelmare commented on tests/components/dynalite/test_light.py:
```diff
@@ -1,44 +1,86 @@
 
-from homeassistant.components.dynalite import DOMAIN
-from homeassistant.components.dynalite.light import DynaliteLight, async_setup_entry
+from asynctest import patch
```
```suggestion
from asynctest import CoroutineMock, patch
```

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -70,24 +63,23 @@
-    bridge = DynaliteBridge(hass, entry)
+    bridge = DynaliteBridge(hass, entry.data[CONF_HOST])
 
     if not await bridge.async_setup():
```
This will never be true.

ziv1234 replied:
fixed

MartinHjelmare commented on homeassistant/components/dynalite/config_flow.py:
```diff
@@ -31,28 +21,20 @@ def __init__(self):
+        self._abort_if_unique_id_configured()
+
+        bridge = DynaliteBridge(self.hass, host)
+        if not await bridge.async_setup():
```
This will never be true.

ziv1234 replied:
fixed this. it now is using try_connection() to try to connect and fails if not successful.

i did leave an override with a config parameter ("nowait"). The reason is that I view Home Assistant as an always-on server. If it reboots for any reason, I want it to be back and fully operational. If there is an event where it reboots and when it comes up the Dynalite gateway is down, I would like it to keep trying. Is there a way to make a component fail to load but have Home Assistant retry after some time? Otherwise, if I wouldn't leave the override, it would never go back to being operational without a manual reboot.

MartinHjelmare replied:
The import flow will not finish on restart if the config entry for this host has been created once. So the scenario you describe won't impact the config flow.

From inside `async_setup_entry` in the component module we can raise `ConfigEntryNotReady` to retry later if the gateway is down.

MartinHjelmare replied:
The retry time is between 5 and 80 seconds:
https://github.com/home-assistant/home-assistant/blob/588f2cd920f955578f97edcab180e1f08d3a14c0/homeassistant/config_entries.py#L223-L241

ziv1234 committed:
pylint fix and removed unnecessary code

ziv1234 committed:
Update tests/components/dynalite/test_light.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
Update tests/components/dynalite/test_light.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
Update tests/components/dynalite/test_light.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
Update tests/components/dynalite/test_light.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
Update tests/components/dynalite/test_light.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
Update tests/components/dynalite/test_light.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
Update tests/components/dynalite/test_light.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
Update tests/components/dynalite/test_light.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
Update tests/components/dynalite/test_light.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
added break in loop

ziv1234 commented on tests/components/dynalite/test_light.py:
```diff
@@ -1,81 +1,84 @@
+    bridge = None
+    for index in hass.data[dynalite.DOMAIN]:
+        if index != dynalite.DATA_CONFIGS:
+            bridge = hass.data[dynalite.DOMAIN][index]
```
doesn't really matter, as there will only be one real bridge, but it looks cleaner and easier to understand so added

ziv1234 committed:
removed last mock_coro reference
pylint fix

MartinHjelmare reviewed (commented):
Please revert the file access mode on hue/test_init.py. The access mode has been changed from 644 to 755.

ziv1234 committed:
added coverage for try_connect

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -70,24 +63,23 @@
 
 async def async_unload_entry(hass, entry):
     """Unload a config entry."""
     LOGGER.error("async_unload_entry %s", entry.data)
```
Please remove this logging or set level to debug.

ziv1234 replied:
sorry. leftover from initial tests. removed

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -52,10 +49,6 @@
         # Store config in hass.data so the config entry can find it
         hass.data[DOMAIN][DATA_CONFIGS][host] = bridge_conf
```
The bridge config schema should be defined in home assistant and not imported from the library and directly used. We should decouple the two APIs, dynalite from home assistant. Please define the bridge validation schema in home assistant.

ziv1234 replied:
ok. will take some work but should be able to get it done later tonight

ziv1234 replied:
done.

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -70,24 +63,23 @@
 async def async_setup_entry(hass, entry):
     """Set up a bridge from a config entry."""
     LOGGER.debug("__init async_setup_entry %s", entry.data)
```
Please start logging messages with capital letter. We don't need to mention what module we're in, as that will be logged by default.

ziv1234 replied:
is this also relevant for debug messages or just INFO and above? Will start cleaning the logs

Also, I don't mention the module. This would be "homeassistant.components.dynalite", right? Just want to debug log the entry point to a specific function if i have to trace stuff

ziv1234 replied:
all logs now capitalized

MartinHjelmare commented on homeassistant/components/dynalite/bridge.py:
```diff
@@ -1,45 +1,32 @@
-        self.all_entities = {}
+        self.async_add_devices = None
+        self.waiting_devices = []
         self.config = None
```
We don't need to store the bridge config as an instance attribute. Just store it in a local variable.

ziv1234 replied:
done

MartinHjelmare commented on tests/components/dynalite/test_bridge.py:
```diff
@@ -1,136 +1,94 @@
+        assert await dyn_bridge.try_connection()
+    # unsuccessful
+    with patch.object(dyn_bridge.dynalite_devices, "connected", False):
+        with patch("asyncio.sleep", return_value=True):
```
Don't patch asyncio. Patch the sleep time and set it to `0`.

To make this easier, extract the sleep time variabel to a module level constant.

ziv1234 replied:
done

MartinHjelmare commented on tests/components/dynalite/test_config_flow.py:
```diff
@@ -1,36 +1,43 @@
+    hass.data[dynalite.DOMAIN][dynalite.DATA_CONFIGS][host] = {dynalite.CONF_HOST: host}
+    with patch.object(dynalite.DynaliteBridge, "async_setup", return_value=setup):
+        with patch.object(
+            dynalite.DynaliteBridge, "try_connection", return_value=connection
```
Please patch the library interface. Don't patch our code.

ziv1234 replied:
can you please explain? Isn't DynaliteBridge the library?

ziv1234 replied:
some additional context - DynaliteBridge is not an Entity in hass
Anyway, please explain and i will fix

ziv1234 replied:
ok. i think i understand what you want. however, i still have a problem. 

i can patch the library so it will return connected or disconnected (like in test_bridge.py). However, if it is disconnected, it will wait 30 seconds to connect, which we don't want in unit-tests. I implemented it and will push shortly, but it still has to patch bridge constants so it won't wait for the full timeout.

ziv1234 replied:
sorry. i thought i had it figured out. the problem is that i need to "edit" the configuration.yaml file so I can supply the config to the component otherwise the bridge doesn't set up. Is there a way to change the config of the hass fixture in the tests?

MartinHjelmare replied:
It's ok to patch the reconnection time constants in the bridge module.

ziv1234 replied:
ok. fixed. thanks for the help!

MartinHjelmare reviewed (commented):
Please patch the library interface instead of our code. That will make the tests more robust.

ziv1234 committed:
added check for a successful connection before bridge.async_setup succeeds
also added a "nowait" config option (default False) that avoids this check

ziv1234 committed:
changed log level

ziv1234 committed:
fixed accidental chmod I did

ziv1234 committed:
fixed accidental change

ziv1234 committed:
not storing config in bridge

ziv1234 committed:
not patching asyncio

ziv1234 commented:
Fixed. Sorry

From: Martin Hjelmare <notifications@github.com>
Sent: Saturday, February 15, 2020 5:32 PM
To: home-assistant/home-assistant <home-assistant@noreply.github.com>
Cc: Ziv <zivk@hotmail.com>; Author <author@noreply.github.com>
Subject: Re: [home-assistant/home-assistant] Enhance Dynalite Integration after review (#31760)


@MartinHjelmare commented on this pull request.

Please revert the file access mode on hue/test_init.py. The access mode has been changed from 644 to 755.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/home-assistant/home-assistant/pull/31760?email_source=notifications&email_token=AD5UNSYISH75XGVQJZDMNGTRDADFNA5CNFSM4KUCYDW2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCVVUFBA#pullrequestreview-359350916>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AD5UNS4RB3GCA2OU56ZP2Y3RDADFNANCNFSM4KUCYDWQ>.

ziv1234 committed:
moved CONFIG_SCHEMA into component

ziv1234 committed:
moved all logs to start capitalized (and revised some of them)

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -8,17 +22,63 @@
+
+
+CHANNEL_DATA_SCHEMA = vol.Schema(
+    vol.All(
```
We don't need `vol.All`. There's only the dict in the schema.

ziv1234 replied:
fixed

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -8,17 +22,63 @@
+    )
+)
+
+CHANNEL_SCHEMA = vol.Schema({num_string: vol.Any(CHANNEL_DATA_SCHEMA, None)})
```
Why should we accept `None` as value?

ziv1234 replied:
fixed

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -8,17 +22,63 @@
+CHANNEL_SCHEMA = vol.Schema({num_string: vol.Any(CHANNEL_DATA_SCHEMA, None)})
+
+AREA_DATA_SCHEMA = vol.Schema(
+    vol.All(
```
Remove `vol.All`.

ziv1234 replied:
fixed

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -8,17 +22,63 @@
+
+BRIDGE_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): str,
```
We prefer to use `cv.string` over just `str`.

ziv1234 replied:
fixed

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -1,5 +1,19 @@
 """Support for the Dynalite networks."""
-from dynalite_devices_lib import BRIDGE_CONFIG_SCHEMA
+from dynalite_devices_lib import (
+    CONF_ACTIVE,
```
It's precarious to use the device library constants for the home assistant config keys. If the library changes we would have to make a breaking change in home assistant. Please decouple the config keys from the library.

ziv1234 replied:
fixed

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -8,17 +22,63 @@
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): str,
+        vol.Required(CONF_HOST): str,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): int,
+        vol.Optional(CONF_LOGLEVEL): str,
```
Log level should not be a config option. Logging level should be changed by using logging filters. This can be done with the `logger` integration in home assistant.

In the library each module should use a logger preferably using the module path:
```py
_LOGGER = logging.getLogger(__name__)
```

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -8,17 +22,63 @@
+        vol.Required(CONF_HOST): str,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): int,
+        vol.Optional(CONF_LOGLEVEL): str,
+        vol.Optional(CONF_AUTODISCOVER, default=False): vol.Coerce(bool),
```
Please separate words with underscore in variable names, eg `CONF_AUTO_DISCOVER`.

ziv1234 replied:
fixed

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -8,17 +22,63 @@
+        vol.Optional(CONF_AREA): AREA_SCHEMA,
+        vol.Optional(CONF_DEFAULT): PLATFORM_DEFAULTS_SCHEMA,
+        vol.Optional(CONF_ACTIVE, default=False): vol.Coerce(bool),
+        vol.Optional(CONF_NOWAIT): vol.Coerce(bool),
```
Please remove the `CONF_NOWAIT` option. It won't be needed.

MartinHjelmare commented on homeassistant/components/dynalite/bridge.py:
```diff
@@ -1,118 +1,89 @@
+        self.waiting_devices = []
+        self.host = host
+        config = hass.data[DOMAIN][DATA_CONFIGS][self.host]
+        self.no_wait = config.get(CONF_NOWAIT, False)
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/dynalite/bridge.py:
```diff
@@ -1,118 +1,89 @@
-        if added_entities:
-            self.add_entities_when_registered(added_entities)
+        if not await self.dynalite_devices.async_setup():
+            return False
```
If we can distinguish between bad set up and the gateway is down, raise two different errors here based on the case. If it's a non recoverable error we want to fail the entry setup. If it's a recoverable error we want to try to set up the entry again later by raising `ConfigEntryNotReady`.

MartinHjelmare commented on homeassistant/components/dynalite/bridge.py:
```diff
@@ -1,118 +1,89 @@
-            self.add_entities_when_registered(added_entities)
+        if not await self.dynalite_devices.async_setup():
+            return False
+        if self.no_wait:
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/dynalite/bridge.py:
```diff
@@ -1,118 +1,89 @@
+        if not await self.dynalite_devices.async_setup():
+            return False
+        if self.no_wait:
+            return True
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/dynalite/const.py:
```diff
@@ -6,6 +6,7 @@
 DATA_CONFIGS = "dynalite_configs"
 
 CONF_BRIDGES = "bridges"
+CONF_NOWAIT = "nowait"
```
```suggestion
```

MartinHjelmare commented on tests/components/dynalite/test_bridge.py:
```diff
@@ -1,136 +1,101 @@
+    return bridge
+
+
+async def test_bridge_setup(dyn_bridge):
```
If we don't patch our bridge code in the init, config flow and light tests, we will already have tested all the bridge code. So then we don't need these extra tests in this module.

ziv1234 replied:
correct. removed them

ziv1234 commented:
Ok. I can change this to ConfigEntryNotReady. What is the timeout that home assistant will retry later (roughly)?

From: Martin Hjelmare <notifications@github.com>
Sent: Saturday, February 15, 2020 9:29 PM
To: home-assistant/home-assistant <home-assistant@noreply.github.com>
Cc: Ziv <zivk@hotmail.com>; Author <author@noreply.github.com>
Subject: Re: [home-assistant/home-assistant] Enhance Dynalite Integration after review (#31760)


@MartinHjelmare commented on this pull request.

________________________________

In homeassistant/components/dynalite/config_flow.py<https://github.com/home-assistant/home-assistant/pull/31760#discussion_r379851059>:

> @@ -31,28 +21,20 @@ def __init__(self):

     async def async_step_import(self, import_info):

         """Import a new bridge as a config entry."""

         LOGGER.debug("async_step_import - %s", import_info)

-        host = self.context[CONF_HOST] = import_info[CONF_HOST]

+        host = import_info[CONF_HOST]

+        await self.async_set_unique_id(host)

+        self._abort_if_unique_id_configured()

+

+        bridge = DynaliteBridge(self.hass, host)

+        if not await bridge.async_setup():

The import flow will not finish on restart if the config entry for this host has been created once. So the scenario you describe won't impact the config flow.

From inside async_setup_entry in the component module we can raise ConfigEntryNotReady to retry later if the gateway is down.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/home-assistant/home-assistant/pull/31760?email_source=notifications&email_token=AD5UNSYZ2JIAYZETNH3LDDTRDA7A7A5CNFSM4KUCYDW2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCVVW45Y#discussion_r379851059>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AD5UNS7FB3HI4XWFM52AFKTRDA7A7ANCNFSM4KUCYDWQ>.

ziv1234 commented:
Ok. Will fix to use this way rather than the override.

The big issue I have is still how to pass the config to hass in the tests. What I put in hass.data[DOMAIN][DATA_CONFIGS] somehow disappears before the DynaliteBridge call. This is in test_config_flow.

Any suggestions?

From: Martin Hjelmare <notifications@github.com>
Sent: Saturday, February 15, 2020 10:03 PM
To: home-assistant/home-assistant <home-assistant@noreply.github.com>
Cc: Ziv <zivk@hotmail.com>; Author <author@noreply.github.com>
Subject: Re: [home-assistant/home-assistant] Enhance Dynalite Integration after review (#31760)


@MartinHjelmare commented on this pull request.

________________________________

In homeassistant/components/dynalite/config_flow.py<https://github.com/home-assistant/home-assistant/pull/31760#discussion_r379852915>:

> @@ -31,28 +21,20 @@ def __init__(self):

     async def async_step_import(self, import_info):

         """Import a new bridge as a config entry."""

         LOGGER.debug("async_step_import - %s", import_info)

-        host = self.context[CONF_HOST] = import_info[CONF_HOST]

+        host = import_info[CONF_HOST]

+        await self.async_set_unique_id(host)

+        self._abort_if_unique_id_configured()

+

+        bridge = DynaliteBridge(self.hass, host)

+        if not await bridge.async_setup():

The retry time is between 5 and 80 seconds:
https://github.com/home-assistant/home-assistant/blob/588f2cd920f955578f97edcab180e1f08d3a14c0/homeassistant/config_entries.py#L223-L241

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/home-assistant/home-assistant/pull/31760?email_source=notifications&email_token=AD5UNS62OTQTIAEUMGDCWADRDBC7JA5CNFSM4KUCYDW2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCVVXJCY#discussion_r379852915>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AD5UNS4Z3CEPWWBIN2KWQMLRDBC7JANCNFSM4KUCYDWQ>.

MartinHjelmare commented on homeassistant/components/dynalite/bridge.py:
```diff
@@ -1,118 +1,89 @@
     """Manages a single Dynalite bridge."""
 
-    def __init__(self, hass, config_entry):
+    def __init__(self, hass, host):
```
Accept the bridge config instead of extracting it from hass.data.

MartinHjelmare commented on homeassistant/components/dynalite/config_flow.py:
```diff
@@ -30,29 +20,21 @@ def __init__(self):
+        await self.async_set_unique_id(host)
+        self._abort_if_unique_id_configured()
+
+        bridge = DynaliteBridge(self.hass, host)
```
Pass the bridge config too.

MartinHjelmare commented on tests/components/dynalite/test_config_flow.py:
```diff
@@ -1,36 +1,43 @@
+    host = "1.2.3.4"
+    hass.data[dynalite.DOMAIN] = {}
+    hass.data[dynalite.DOMAIN][dynalite.DATA_CONFIGS] = {}
+    hass.data[dynalite.DOMAIN][dynalite.DATA_CONFIGS][host] = {dynalite.CONF_HOST: host}
```
Remove all of this hass.data modification. The config flow should be isolated from the rest of the integration. It should only accept the data that we can pass when starting the flow.

ziv1234 committed:
moved test_config_flow to not patch the DynaliteBridge

ziv1234 committed:
also took DynaliteBridge patching out of test_init

ziv1234 committed:
removed NO_WAIT

MartinHjelmare commented on homeassistant/components/dynalite/bridge.py:
```diff
@@ -1,118 +1,83 @@
             updateDeviceFunc=self.update_device,
         )
 
     async def async_setup(self, tries=0):
```
`tries` is never used.

ziv1234 replied:
fixed

ziv1234 committed:
fixes to SCHEMA

ziv1234 committed:
changed _ in multi-word CONF
moved imports to component const.py

ziv1234 committed:
removed tries

ziv1234 committed:
removed redundant tests

ziv1234 committed:
fixed some small change i broke in the library. only version update

ziv1234 committed:
fixed rewuirements

MartinHjelmare commented on tests/components/dynalite/test_light.py:
```diff
@@ -1,44 +1,78 @@
+    with patch(
+        "homeassistant.components.dynalite.bridge.DynaliteDevices.async_setup",
+        return_value=True,
+    ), patch.object(dynalite.DynaliteBridge, "try_connection", return_value=True):
```
```suggestion
    ), patch("homeassistant.components.dynalite.bridge.DynaliteDevices.available", True):
```

MartinHjelmare commented on tests/components/dynalite/test_config_flow.py:
```diff
@@ -1,36 +1,41 @@
+    """Run a flow with or without errors and return result."""
+    host = "1.2.3.4"
+    with patch(
+        "dynalite_devices_lib.DynaliteDevices.async_setup", return_value=setup
```
```suggestion
        "homeassistant.components.dynalite.bridge.DynaliteDevices.async_setup", return_value=setup
```

MartinHjelmare commented on tests/components/dynalite/test_config_flow.py:
```diff
@@ -1,36 +1,41 @@
+    host = "1.2.3.4"
+    with patch(
+        "dynalite_devices_lib.DynaliteDevices.async_setup", return_value=setup
+    ), patch("dynalite_devices_lib.DynaliteDevices.available", connection), patch(
```
```suggestion
    ), patch("homeassistant.components.dynalite.bridge.DynaliteDevices.available", connection), patch(
```

MartinHjelmare commented on homeassistant/components/dynalite/bridge.py:
```diff
@@ -1,118 +1,83 @@
+
 from dynalite_devices_lib import DynaliteDevices
-from dynalite_lib import CONF_ALL
+from dynalite_lib import CONF_ALL, CONF_HOST
```
We should use the constant we have defined to access the config we have created. These constants are library specific and should just be used when passing an item to a library interface.

ziv1234 replied:
fixed

ziv1234 replied:
fixed

MartinHjelmare commented on homeassistant/components/dynalite/__init__.py:
```diff
@@ -1,24 +1,81 @@
+    {
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Optional(CONF_FADE): vol.Coerce(float),
+        vol.Optional(CONF_HIDDEN_ENTITY, default=False): vol.Coerce(bool),
```
Please remove this option. The entity.hidden property is deprecated and should not be used.

ziv1234 replied:
removed

MartinHjelmare reviewed (commented):
Please remove the light entity `hidden` property and set `device_info` specifically in the light entity. The `device_info` property is a home assistant interface and should not be directly coupled to the dynalite interface.

ziv1234 committed:
Update tests/components/dynalite/test_config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
Update tests/components/dynalite/test_light.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
Update tests/components/dynalite/test_config_flow.py

Co-Authored-By: Martin Hjelmare <marhje52@gmail.com>

ziv1234 committed:
removed HIDDEN_ENTITY
removed hass in test fixture

ziv1234 committed:
Merge branch 'dev' of https://github.com/ziv1234/home-assistant into dev

ziv1234 commented on tests/components/dynalite/test_light.py:
```diff
@@ -1,44 +1,78 @@
-async def test_light():
-    """Test the light entity."""
+@pytest.fixture
+def mock_device(hass):
```
fixed

ziv1234 commented:
done

From: Martin Hjelmare <notifications@github.com>
Sent: Monday, February 17, 2020 4:51 PM
To: home-assistant/home-assistant <home-assistant@noreply.github.com>
Cc: Ziv <zivk@hotmail.com>; Author <author@noreply.github.com>
Subject: Re: [home-assistant/home-assistant] Enhance Dynalite Integration after review (#31760)


@MartinHjelmare commented on this pull request.

Please remove the light entity hidden property and set device_info specifically in the light entity. The device_info property is a home assistant interface and should not be directly coupled to the dynalite interface.

________________________________

In tests/components/dynalite/test_light.py<https://github.com/home-assistant/home-assistant/pull/31760#discussion_r380191783>:

> -async def test_light_setup():



-    """Test a successful setup."""



-    hass = Mock()



-    entry = Mock()



-    async_add = Mock()



-    bridge = Mock()



-    hass.data = {DOMAIN: {entry.entry_id: bridge}}



-    await async_setup_entry(hass, entry, async_add)



-    bridge.register_add_entities.assert_called_once()



-    assert bridge.register_add_entities.mock_calls[0] == call(async_add)



-



-



-async def test_light():



-    """Test the light entity."""



+@pytest.fixture<mailto:+@pytest.fixture>



+def mock_device(hass):



hass isn't used.
⬇️ Suggested change

-def mock_device(hass):



+def mock_device():



________________________________

In tests/components/dynalite/test_light.py<https://github.com/home-assistant/home-assistant/pull/31760#discussion_r380200461>:

> +    device.name = "NAME"



+    device.device_info = {



+        "identifiers": {(dynalite.DOMAIN, device.unique_id)},



+        "name": device.name,



+        "manufacturer": "Dynalite",



+    }



+    return device



+



+



+async def create_light_from_device(hass, device):



+    """Set up the component and platform and create a light based on the device provided."""



+    host = "1.2.3.4"



+    with patch(



+        "homeassistant.components.dynalite.bridge.DynaliteDevices.async_setup",



+        return_value=True,



+    ), patch.object(dynalite.DynaliteBridge, "try_connection", return_value=True):


⬇️ Suggested change

-    ), patch.object(dynalite.DynaliteBridge, "try_connection", return_value=True):



+    ), patch("homeassistant.components.dynalite.bridge.DynaliteDevices.available", True):



________________________________

In tests/components/dynalite/test_config_flow.py<https://github.com/home-assistant/home-assistant/pull/31760#discussion_r380202139>:

> -                )



-                mock_create.assert_called_once()



-                assert mock_create.mock_calls[0] == call(



-                    title=host, data={"host": host}



-                )



+from asynctest import patch



+



+from homeassistant import config_entries



+from homeassistant.components import dynalite



+



+



+async def run_flow(hass, setup, connection):



+    """Run a flow with or without errors and return result."""



+    host = "1.2.3.4"



+    with patch(



+        "dynalite_devices_lib.DynaliteDevices.async_setup", return_value=setup


⬇️ Suggested change

-        "dynalite_devices_lib.DynaliteDevices.async_setup", return_value=setup



+        "homeassistant.components.dynalite.bridge.DynaliteDevices.async_setup", return_value=setup



________________________________

In tests/components/dynalite/test_config_flow.py<https://github.com/home-assistant/home-assistant/pull/31760#discussion_r380202591>:

> -                mock_create.assert_called_once()



-                assert mock_create.mock_calls[0] == call(



-                    title=host, data={"host": host}



-                )



+from asynctest import patch



+



+from homeassistant import config_entries



+from homeassistant.components import dynalite



+



+



+async def run_flow(hass, setup, connection):



+    """Run a flow with or without errors and return result."""



+    host = "1.2.3.4"



+    with patch(



+        "dynalite_devices_lib.DynaliteDevices.async_setup", return_value=setup



+    ), patch("dynalite_devices_lib.DynaliteDevices.available", connection), patch(


⬇️ Suggested change

-    ), patch("dynalite_devices_lib.DynaliteDevices.available", connection), patch(



+    ), patch("homeassistant.components.dynalite.bridge.DynaliteDevices.available", connection), patch(



________________________________

In homeassistant/components/dynalite/bridge.py<https://github.com/home-assistant/home-assistant/pull/31760#discussion_r380207089>:

>  from dynalite_devices_lib import DynaliteDevices



-from dynalite_lib import CONF_ALL



+from dynalite_lib import CONF_ALL, CONF_HOST



We should use the constant we have defined to access the config we have created. These constants are library specific and should just be used when passing an item to a library interface.

________________________________

In homeassistant/components/dynalite/__init__.py<https://github.com/home-assistant/home-assistant/pull/31760#discussion_r380223372>:

> +)



+



+



+def num_string(value):



+    """Test if value is a string of digits, aka an integer."""



+    new_value = str(value)



+    if new_value.isdigit():



+        return new_value



+    raise vol.Invalid("Not a string with numbers")



+



+



+CHANNEL_DATA_SCHEMA = vol.Schema(



+    {



+        vol.Optional(CONF_NAME): cv.string,



+        vol.Optional(CONF_FADE): vol.Coerce(float),



+        vol.Optional(CONF_HIDDEN_ENTITY, default=False): vol.Coerce(bool),



Please remove this option. The entity.hidden property is deprecated and should not be used.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/home-assistant/home-assistant/pull/31760?email_source=notifications&email_token=AD5UNS27C2ETSLHTIWGPDVLRDKP4ZA5CNFSM4KUCYDW2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCVYUX6Q#pullrequestreview-359746554>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AD5UNS62ZLPLQNROAEUAWV3RDKP4ZANCNFSM4KUCYDWQ>.

ziv1234 committed:
black fixes

ziv1234 committed:
removed final piece of hidden_entity from light
fix in the library
updated config flow so if the entry is already set but with a different config, calls async_update_entry

ziv1234 committed:
removed DATA_CONFIGS - no longer necessary

ziv1234 committed:
pylint fixes

ziv1234 committed:
added coverage

ziv1234 commented:
ok. all the comments should be implemented now.

One thing that I am not sure about is in the implementation in config_flow.py. When I was simply using _abort_if... it had the problem that if I would change configuration.yaml, it wouldn't update the devices in home assistant because it wouldn't reconfigure. I tried removing the old entry and creating a new one, and it sort of works. The problem is that when I change the configuration.yaml and restart HA, it still runs with the old config. When I restart HA again, it now loads the new config. Is there a better way to do this?

MartinHjelmare commented:
If config yaml items change besides the unique id item (host), we can update the existing entry in the config flow before aborting.

Pass the updates in `self._abort_if_unique_id_configured`.

https://github.com/home-assistant/home-assistant/blob/f75bda75512e36faaebddcd314149553d99d9f6f/homeassistant/config_entries.py#L820-L833

MartinHjelmare commented:
A side note: Answering comments via email is unfortunately very spammy and unreadable in the PR thread. Try to avoid that and answer via github web instead.

MartinHjelmare commented on tests/components/dynalite/test_config_flow.py:
```diff
@@ -1,36 +1,91 @@
+    ), patch(
+        "homeassistant.components.dynalite.bridge.DynaliteDevices.available", True
+    ):
+        await hass.config_entries.flow.async_init(
```
Add the first entry as a `MockConfigEntry` instead and use entry.add_to_hass.

https://github.com/home-assistant/home-assistant/blob/f75bda75512e36faaebddcd314149553d99d9f6f/tests/common.py#L656-L694

ziv1234 replied:
done

ziv1234 commented:
> If config yaml items change besides the unique id item (host), we can update the existing entry in the config flow before aborting.
> 
> Pass the updates in `self._abort_if_unique_id_configured`.
> 
> https://github.com/home-assistant/home-assistant/blob/f75bda75512e36faaebddcd314149553d99d9f6f/homeassistant/config_entries.py#L820-L833

The problem is that it only updates if there are more items there. Perhaps better if I reintroduce DATA_CONFIGS but will do it more correctly this time. Is this a standard way? I saw that hue component uses it but not sure if it is common

ziv1234 commented:
> A side note: Answering comments via email is unfortunately very spammy and unreadable in the PR thread. Try to avoid that and answer via github web instead.

sure. will do

MartinHjelmare commented:
> The problem is that it only updates if there are more items there.

What do you mean?

ziv1234 commented:
> > The problem is that it only updates if there are more items there.
> 
> What do you mean?

not updates.items() <= entry.data.items()

wouldn't it only work if there are more items in the new config?

Anyway, if the DATA_CONFIGS pattern is ok, I can easily implement it correctly. Will do so now, so please let me know whether or not to push

MartinHjelmare commented:
I think that will just check if the existing data is the same as the updates and not update if it's the same.

I think we should use my suggested approach and not do anything extra.

ziv1234 committed:
use abort in config_flow

ziv1234 commented:
> I think that will just check if the existing data is the same as the updates and not update if it's the same.
> 
> I think we should use my suggested approach and not do anything extra.

sure. done

MartinHjelmare commented on tests/components/dynalite/test_config_flow.py:
```diff
@@ -1,36 +1,67 @@
+
+
+async def test_existing(hass):
+    """Test when the entry exists with the same config."""
```
It's good to add a test that tests that updates are applied on an existing entry with the same host.

ziv1234 replied:
i added a test, but not sure how to check the update

MartinHjelmare replied:
Check the data of the mock entry. It should be updated with the new items.

ziv1234 replied:
fixed

ziv1234 committed:
test update

ziv1234 committed:
removed logs

ziv1234 committed:
test that update actually updates the entry

MartinHjelmare reviewed (approved):
Looks good! Nice work!

ziv1234 commented:
> Looks good! Nice work!

Thanks! Appreciate all the work you did. Learned a lot about Home Assistant.

Now I can add the switch and cover platforms in another pull request after this one is merged

frenck commented:
The host isn't a unique ID? It can change...
So you are using host as the unique ID and want to update it with the same thing that is the unique ID?

IMHO, a unique ID should not be used here.

frenck commented on homeassistant/components/dynalite/config_flow.py:
```diff
@@ -30,29 +20,16 @@ def __init__(self):
-        return self.async_create_entry(title=host, data={CONF_HOST: host})
+        LOGGER.debug("Starting async_step_import - %s", import_info)
+        host = import_info[CONF_HOST]
+        await self.async_set_unique_id(host)
```
The host can change, and should not be used as a unique ID here.

MartinHjelmare commented:
This is just a unique id for the config entry not the entities. The config entry would need to be replaced anyway if the host would change. Otherwise the integration would stop working.

ziv1234 commented:
The host is the most constant thing. This is the IP of the gateway. There is more info in the config entry so I update it with the same host and the rest of the data. You can see what I put on the issue I opened. Is there another place in has to store persistent data?

frenck commented:
I've seen your issue @ziv1234.

I think this has mainly to do with existing entries, that don't have a unique ID registered yet, but are in the registry.

You can solve that by updating the entry on entry setup, like so:

https://github.com/home-assistant/home-assistant/blob/dd8597cb46c5e788995e1ea56b87dfbaf43ca51c/homeassistant/components/wled/__init__.py#L64-L68

ziv1234 commented:
maybe i misunderstood something, but it seems like they are handling something else. In wled they are looking at an entry that doesn't have unique_id so they configure one for it. In my case, all entries will have unique_id, but they will be running with a different data (entry.data) which is what I expected to update.

In dynalite, I use entry.data["host"] as the unique_id, but the other parameters (port, areas / devices, etc.) could change, so would expect to keep the unique_id and update it.

It works, but the problem is that it requires and extra restart of HA.

Is there something I should do differently here?

MartinHjelmare commented:
The config entry will be updated immediately before aborting the config flow when checking for unique config entry id.

If you need to take some action in the integration when the config entry data is updated, I think you have to register a listener.

https://developers.home-assistant.io/docs/config_entries_options_flow_handler#signal-updates

I think this listener is required if the config entry already has been loaded when the update happens.

MartinHjelmare commented:
Marking this for 0.106 since the first PR for this integration was merged before beta cut and this PR is cleaning up many things about the first PR.

ziv1234 commented:
ok. will start thinking about how to implement the signal update one. It makes sense.

Is there a way to ask home assistant to unload and reload an entry?

MartinHjelmare commented:
Yes, `hass.config_entries.async_reload`.

https://github.com/home-assistant/home-assistant/blob/2fb66fd866b39f210f48edb14d3e8b21f62ff974/homeassistant/config_entries.py#L705

MartinHjelmare commented:
Can we merge here now, and further changes can go in a future PR?

MartinHjelmare commented:
Btw, we need to update the docs, since we removed some config items. Please link a docs PR in the PR description. Then we can merge.

ziv1234 commented:
done. created PR #12157

ziv1234 commented:
we can merge now. i believe that the change to use reload_entry could be tricky, and the problem only happens when changing configuration.yaml, which shouldn't happen often after the initial setup
